### convert-SURVIVOR-to-circlize.r ####################################################################
# This script converts the output VCF file containing consensus VCFs generated by SURVIVOR for
# structural variants into a format suitable for the circlize R package.

# Authors: Raag Agrawal (2023-04-03), Selina Wu (2023-05-08)

### NOTES ##########################################################################################
# The input VCF file should be generated separately and prior to circos plotting. User will need to
# have downloaded SURVIVOR (https://github.com/fritzsedlazeck/SURVIVOR) to find consensus VCFs.


### PREAMBLE #######################################################################################
library(circlize);
library(vcfR);

date <- Sys.Date();


### FUNCTIONS ######################################################################################
## Function: convert.survivor.to.circlize ---------------------------------------------------------
# Description:
    # Function that reads in the input VCF file and extracts relevant SV data
# Input variables:
    # input.vcf (character) - The input VCF file
    # sample.id (character) - Sample name
# Output variables:
    # sv.data.survivor (data.frame) Data frame containing structural variant data

convert.survivor.to.circlize <- function(vcf, sample.id) {
    # Extract data from the INFO column
    s.info <- INFO2df(vcf);

    # Extract data from s.info
    type <- s.info$SVTYPE;
    sample <- rep(sample.id, length(type));

    # Start chromosome, start position, and end position
    chr.start <- getCHROM(vcf);
    start <- getPOS(vcf);
    chr.end <- s.info$CHR2;
    end <- s.info$END;
    alt <- getALT(vcf);

    # Save extracted data in df
    sv.data.survivor <- data.frame(
        sample = sample,
        chr.start = chr.start,
        start = start,
        chr.end = chr.end,
        end = end,
        type = type,
        alt = alt
        );

    return(sv.data.survivor);
    }

### FIN ############################################################################################
