### convert-manta-to-circlize.r ####################################################################
# This script converts the output VCF file generated by Manta for structural variants into a format
# suitable for the circlize R package.

# Authors: Raag Agrawal (2023-04-03), Selina Wu (2023-05-08)

### NOTES ##########################################################################################
# The input VCF file should have been generated using Manta for structural variants detection.

### PREAMBLE #######################################################################################
library(circlize);
library(vcfR);

date <- Sys.Date();

### FUNCTIONS ######################################################################################
## Function: convert.manta.sv.to.circlize ---------------------------------------------------------
# Description:
    # Function that reads in the input VCF file and extracts relevant BND (including
    # interchromosomal translocations) data
# Input variables:
    # input.vcf (character) - The input VCF file
    # sample.id (character) - Sample name
# Output variables:
    # sv.data.manta (data.frame) Data frame containing structural variant data

convert.manta.BND.to.circlize <- function(vcf, sample.id) {
    # Extract data from the INFO column
    s.info <- INFO2df(vcf);

    # Extract data from s.info
    type <- s.info$SVTYPE;
    sample <- rep(sample.id, length(type));

    # Start chromosome, start position, and end position
    chr.start <- getCHROM(vcf);
    start <- getPOS(vcf);
    # For BNDs, set end chromosome and position to those indicated in 'ALT'
    alt <- getALT(vcf);
    chr.end <- gsub("^.*?(chr\\w+).*", "\\1", alt); # nolint
    end <- gsub("^.*:(\\d+).*$", "\\1", alt); # nolint

    # Save extracted data in df
    sv.data.manta <- data.frame(
        sample = sample,
        chr.start = chr.start,
        start = start,
        chr.end = chr.end,
        end = end,
        type = type,
        alt = alt
        );

    return(sv.data.manta);
    }


## Function: convert.manta.otherSV.to.circlize -----------------------------------------------------
# Description:
    # Function that reads in the input VCF file and extracts relevant INS, DEL, DUP data
# Input variables:
    # input.vcf (character) - The input VCF file
    # sample.id (character) - Sample name
# Output variables:
    # sv.data.manta (data.frame) Data frame containing structural variant data

convert.manta.otherSV.to.circlize <- function(vcf, sample.id) {
    # Extract data from the INFO column
    s.info <- INFO2df(vcf);

    # Extract data from s.info
    type <- s.info$SVTYPE;
    sample <- rep(sample.id, length(type));
    alt <- getALT(vcf);

    # Start chromosome, start position, and end position
    chr.start <- getCHROM(vcf);
    start <- getPOS(vcf);
    # For other SVs, set end chromosome to start chr
    chr.end <- chr.start;
    end <- s.info$END;

    # Save extracted data in df
    sv.data.manta <- data.frame(
        sample = sample,
        chr.start = chr.start,
        start = start,
        chr.end = chr.end,
        end = end,
        type = type,
        alt = alt
        );

    return(sv.data.manta);
    }

### FIN ############################################################################################
