### convert-delly-to-circlize.r ####################################################################
# This script converts the output VCF file generated by DELLY for structural variants into a format
# suitable for the circlize R package.

# Authors: Raag Agrawal (2023-04-03), Selina Wu (2023-05-08)

### NOTES ##########################################################################################
# The input VCF file should have been generated using DELLY for structural variants detection.

### PREAMBLE #######################################################################################
library(circlize);
library(vcfR);

date <- Sys.Date();


### FUNCTIONS ######################################################################################
## Function: convert.delly.BND.to.circlize ---------------------------------------------------------
# Description:
    # Function that reads in the input VCF file and extracts relevant BND (including
    # interchromosomal translocations) data
# Input variables:
    # input.vcf (character) - The input VCF file
    # sample (character) - Sample name
# Output variables:
    # sv.data (data.frame) Data frame containing structural variant data

convert.delly.BND.circlize <- function(vcf, sample) {
    # Extract INFO data
    s.info <- INFO2df(vcf);

    # Extract s.info data
    type <- s.info$SVTYPE;
    sample <- rep(sample, length(type));

    # Start chromosome, start position, and end position
    chr.start <- getCHROM(vcf);
    start <- getPOS(vcf);
    # For BND, the ending chromosome and position will be data stored in 'CHR2' and position in 'POS2', respectively
    end <- s.info$POS2;
    chr.end <- s.info$CHR2;

    alt <- getALT(vcf);

    # Save extracted info as a df
    sv.data <- data.frame(
        sample = sample,
        chr.start = chr.start,
        start = start,
        chr.end = chr.end,
        end = end,
        type = type,
        alt = alt
        );

    return(sv.data);
    }

## Function: convert.delly.otherSV.to.circlize ---------------------------------------------------------
# Description:
    # Function that reads in the input VCF file and extracts relevant INS, INV, DEL, DUP data
# Input variables:
    # input.vcf (character) - The input VCF file
    # sample (character) - Sample name
# Output variables:
    # sv.data (data.frame) Data frame containing structural variant data

convert.delly.otherSV.to.circlize <- function(vcf, sample) {
    # Extract INFO data
    s.info <- INFO2df(vcf);

    # Extract s.info data
    type <- s.info$SVTYPE;
    sample <- rep(sample, length(type));

    # Start chromosome, start position, and end position
    chr.start <- getCHROM(vcf);
    start <- getPOS(vcf);
    # For SVs that are not BND, ending chromosome will be the same as starting chromosome
    chr.end <- chr.start
    end <- s.info$END;

    alt <- getALT(vcf);

    # Save extracted info as a df
    sv.data <- data.frame(
        sample = sample,
        chr.start = chr.start,
        start = start,
        chr.end = chr.end,
        end = end,
        type = type,
        alt = alt
        );

    return(sv.data);
    }

### FIN ############################################################################################
